_posts/2018-10-03-ssl-logging-in-otp-22.md

```diff
@@ -0,0 +1,268 @@
+      <<0,159>>,
+      <<0,163>>,
+      <<0,107>>,
+      <<0,106>>,
+      <<"À+">>,<<"À/">>,<<"À#">>,<<"À'">>,<<"À-">>,<<"À1">>,<<"À%">>,<<"À)">>,
+      <<0,158>>,
+      <<0,162>>,
+      <<0,103>>,
+      <<0,64>>,
+      <<"À\n">>,
+      <<192,20>>,
+      <<0,57>>,
+      <<0,56>>,
+      <<192,5>>,
+      <<192,15>>,
+      <<"À\t">>,
+      <<192,19>>,
+      <<0,51>>,
+      <<0,50>>,
+      <<192,4>>,
+      <<192,14>>]},
+ {compression_methods,[0]},
+...
+[Truncated for brevity]
+```
+
+This is not the final format as there are many ways to further improve the representation
+of the handshake protocol messages such as converting the cipher suites to a human readable
+erlang representation.
+
+As a comparison this is the debug output from an OpenSSL server when the same erlang client
+connects to it:
+
+
+```
+$ /usr/bin/openssl s_server -accept 11000 -tls1_2 -cert server.pem -key server.key -msg -debug
+Using default temp DH parameters
+ACCEPT
+read from 0x16f0040 [0x16f56b3] (5 bytes => 5 (0x5))
+0000 - 16 03 03 00 a1                                    .....
+<<< ??? [length 0005]
+    16 03 03 00 a1
+read from 0x16f0040 [0x16f56b8] (161 bytes => 161 (0xA1))
+0000 - 01 00 00 9d 03 03 5b ac-a1 cc 20 4c 4d 52 d0 d4   ......[... LMR..
+0010 - c8 fc dd 95 b0 fa 65 97-57 9e 44 aa dd 0e 46 10   ......e.W.D...F.
+0020 - 6c 14 57 9c ce a0 00 00-04 00 ff c0 14 01 00 00   l.W.............
+0030 - 70 00 2b 00 06 00 04 03-04 03 03 00 00 00 0e 00   p.+.............
+0040 - 0c 00 00 09 6c 6f 63 61-6c 68 6f 73 74 00 0a 00   ....localhost...
+0050 - 3a 00 38 00 0e 00 0d 00-19 00 1c 00 0b 00 0c 00   :.8.............
+0060 - 1b 00 18 00 09 00 0a 00-1a 00 16 00 17 00 08 00   ................
+0070 - 06 00 07 00 14 00 15 00-04 00 05 00 12 00 13 00   ................
+0080 - 01 00 02 00 03 00 0f 00-10 00 11 00 0b 00 02 01   ................
+0090 - 00 00 32 00 04 00 02 02-03 00 0d 00 04 00 02 02   ..2.............
+00a0 - 01                                                .
+<<< TLS 1.2 Handshake [length 00a1], ClientHello
+    01 00 00 9d 03 03 5b ac a1 cc 20 4c 4d 52 d0 d4
+    c8 fc dd 95 b0 fa 65 97 57 9e 44 aa dd 0e 46 10
+    6c 14 57 9c ce a0 00 00 04 00 ff c0 14 01 00 00
+    70 00 2b 00 06 00 04 03 04 03 03 00 00 00 0e 00
+    0c 00 00 09 6c 6f 63 61 6c 68 6f 73 74 00 0a 00
+    3a 00 38 00 0e 00 0d 00 19 00 1c 00 0b 00 0c 00
+    1b 00 18 00 09 00 0a 00 1a 00 16 00 17 00 08 00
+    06 00 07 00 14 00 15 00 04 00 05 00 12 00 13 00
+    01 00 02 00 03 00 0f 00 10 00 11 00 0b 00 02 01
+    00 00 32 00 04 00 02 02 03 00 0d 00 04 00 02 02
+    01
+...
+[Truncated for brevity]
+```
+
+The verbose debug logging proved to be especially useful during the development of new
+extensions as previously we had to use wireshark captures to validate TLS protocol
+messages.
+
+## Implementation
+
+In the ```ssl``` application we needed a way to handle two types of protocol
+messages (tls_record and handshake) each with a custom formatter function.
```

Comment:
messages, tld_record and handshake, each

Reply:
Will fix.